src/nupic/encoders/RandomDistributedScalarEncoder.cpp

```diff
@@ -0,0 +1,149 @@
+ * --------------------------------------------------------------------- */
+
+/** @file
+ * Implementation of the RandomDistributedScalarEncoder
+ */
+
+#include <nupic/encoders/RandomDistributedScalarEncoder.hpp>
+#include <nupic/utils/MurmurHash3.hpp>
+#include <nupic/utils/Random.hpp>
+#include <algorithm> // fill
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::encoders;
+
+RandomDistributedScalarEncoder::RandomDistributedScalarEncoder(
+                                              const RDSE_Parameters &parameters)
+  { initialize( parameters ); }
+
+void RandomDistributedScalarEncoder::initialize( const RDSE_Parameters &parameters)
+{
+  // Check size parameter
+  NTA_CHECK( parameters.size > 0u );
+
+  // Initialize parent class.
+  BaseEncoder<Real64>::initialize({ parameters.size });
+
+  // Check other parameters
+  UInt num_active_args = 0;
+  if( parameters.activeBits > 0u)   { num_active_args++; }
+  if( parameters.sparsity   > 0.0f) { num_active_args++; }
+  NTA_CHECK( num_active_args != 0u )
+      << "Missing argument, need one of: 'activeBits' or 'sparsity'.";
+  NTA_CHECK( num_active_args == 1u )
+      << "Too many arguments, choose only one of: 'activeBits' or 'sparsity'.";
+
+  UInt num_resolution_args = 0;
+  if( parameters.radius     > 0.0f) { num_resolution_args++; }
+  if( parameters.resolution > 0.0f) { num_resolution_args++; }
+  NTA_CHECK( num_resolution_args != 0u )
+      << "Missing argument, need one of: 'radius', 'resolution'.";
+  NTA_CHECK( num_resolution_args == 1u )
+      << "Too many arguments, choose only one of: 'radius', 'resolution'.";
+
+  args_ = parameters;
+  // Finish filling in all of parameters.
+
+  // Determine number of activeBits.
+  if( args_.sparsity > 0.0f ) {
+    NTA_CHECK( args_.sparsity >= 0.0f );
+    NTA_CHECK( args_.sparsity <= 1.0f );
+    args_.activeBits = (UInt) round( args_.size * args_.sparsity );
+    NTA_CHECK( args_.activeBits > 0u );
+  }
+  // Determine sparsity. Always calculate this even if it was given, to correct for rounding error.
+  args_.sparsity = (Real) args_.activeBits / args_.size;
+
+  // Determine resolution.
+  if( args_.radius > 0.0f ) {
+    args_.resolution = args_.radius / args_.activeBits;
+  }
+  // Determine radius.
+  else if( args_.resolution > 0.0f ) {
+    args_.radius = args_.activeBits * args_.resolution;
+  }
+
+  while( args_.seed == 0u ) {
+    args_.seed = Random().getUInt32();
+  }
+}
+
+void RandomDistributedScalarEncoder::encode(Real64 input, sdr::SDR &output)
+{
+  // Check inputs
+  NTA_CHECK( output.size == size );
+  if( isnan(input) ) {
+    output.zero();
+    return;
+  }
+
+  auto &data = output.getDense();
+  fill( data.begin(), data.end(), 0u );
+
+  const UInt index = (UInt) (input / args_.resolution);
+  for(auto offset = 0u; offset < args_.activeBits; ++offset)
+  {
+    UInt hash_buffer = index + offset;
+    UInt32 bucket = MurmurHash3_x86_32(&hash_buffer, sizeof(hash_buffer), args_.seed);
```

Comment:
is this correct? &hash_buffer is an address of int

Reply:
This is correct. The hash function accepts a void* data buffer and the sizeof the buffer.